### PR TITLE
Enhancement: Added interface for GMSD and Multi-Scale GMSD

### PR DESCRIPTION
**Pull request by denproc**
## Proposed Changes
- Added interface for GMSD and Multi-Scale GMSD to use them as functions without creating the Layer (also to match other metrics, which are available as function and as nn.Layer);
- Optimised colour conversion;
- Optimised tests, checked on cuda.

### CODE REVIEW

File: piq/gmsd.py [hunk 1]
```diff
@@ -29,25 +29,59 @@ def _gmsd(prediction: torch.Tensor, target: torch.Tensor,
...
+    References:
+        https://arxiv.org/pdf/1308.3052.pdf
+    """
+
+    _validate_input(input_tensors=(prediction, target), allow_5d=False, scale_weights=None)
+    prediction, target = _adjust_dimensions(input_tensors=(prediction, target))
+
+    if data_range == 255:
+        prediction = prediction / 255.
+        target = target / 255.
```
Comment (zakajd) [inline piq/gmsd.py:48]: Let's to range conversion the same way it's done in VSI and other more recent metrics.
```suggestion
    prediction = prediction / float(data_range)
    target = target / float(data_range)
```

File: piq/gmsd.py [hunk 2]
```diff
@@ -107,27 +135,99 @@ def forward(self, prediction: torch.Tensor, target: torch.Tensor) -> torch.Tenso
...
+    # Check that input is big enough
+    num_scales = scale_weights.size(0)
+    min_size = 2 ** num_scales + 1
+
+    if prediction.size(-1) < min_size or prediction.size(-2) < min_size:
+        raise ValueError(f'Invalid size of the input images, expected at least {min_size}x{min_size}.')
+
+    if data_range == 255:
+        prediction = prediction / 255.
+        target = target / 255.
```
Comment (zakajd) [inline piq/gmsd.py:189]: Same as above
```suggestion
    prediction = prediction / float(data_range)
    target = target / float(data_range)
```

File: piq/gmsd.py [hunk 3]
```diff
@@ -165,12 +265,14 @@ def __init__(self, reduction: str = 'mean', data_range: Union[int, float] = 1.,
...
-        
+
         # Values from the paper
         if scale_weights is None:
             self.scale_weights = torch.tensor([0.096, 0.596, 0.289, 0.019])
-        else:
+        elif isinstance(scale_weights, torch.Tensor):
             # Normalize scale weights
+            self.scale_weights = scale_weights / scale_weights.sum()
+        else:
```
Comment (zakajd) [inline piq/gmsd.py:275]: You already have this check inside `multi_scale_gmsd` function. Propose not duplicate it here. 
Just save scale_weights:
`self.scale_weights = scale_weights`

File: piq/gmsd.py [hunk 4]
```diff
@@ -179,70 +281,17 @@ def __init__(self, reduction: str = 'mean', data_range: Union[int, float] = 1.,
...
         Args:
-            prediction: Tensor of prediction of the network.
-            target: Reference tensor.
+            prediction: Tensor of prediction of the network. Required to be
+                2D (H, W), 3D (C,H,W) or 4D (N,C,H,W), channels first.
+            target: Reference tensor. Required to be 2D (H, W), 3D (C,H,W) or 4D (N,C,H,W), channels first.
 
         Returns:
-            Value of GMSD loss to be minimized. 0 <= GMSD loss <= 1.
+            Value of SSIM loss to be minimized, i.e 1 - `ssim`. 0 <= SSIM loss <= 1. In case of 5D input tensors,
```
Comment (zakajd) [inline piq/gmsd.py:292]: Do we really support complex tensors? 
```suggestion
            Value of GMSD loss to be minimized. 0 <= GSMD loss <= 1. In case of 5D input tensors,
```

Comment (denproc) [reply piq/gmsd.py:292]: Nope

File: tests/test_gmsd.py [hunk 5]
```diff
@@ -16,160 +18,242 @@ def target() -> torch.Tensor:
...
+    measure = gmsd(prediction.to(device), target.to(device))
+
+    measure_255 = gmsd(prediction_255.to(device), target_255.to(device), data_range=255)
+    diff = torch.abs(measure_255 - measure)
+    assert diff <= 1e-4, f'Result for same tensor with different data_range should be the same, got {diff}'
+
+
+def test_gmsd_supports_greyscale_tensors(device: str):
+    target = torch.ones(2, 1, 128, 128)
+    prediction = torch.zeros(2, 1, 128, 128)
```
Comment (snk4tr) [inline tests/test_gmsd.py:70]: Consider reducing spacial resolution of tensors for the sake of speed here

File: tests/test_gmsd.py [hunk 6]
```diff
@@ -16,160 +18,242 @@ def target() -> torch.Tensor:
...
     prediction.requires_grad_()
-    loss_value = GMSDLoss()(prediction.cuda(), target.cuda())
+    loss_value = GMSDLoss()(prediction.to(device), target.to(device))
     loss_value.backward()
-    assert prediction.grad is not None, LEAF_VARIABLE_ERROR_MESSAGE
+    assert torch.isfinite(prediction.grad).all(), LEAF_VARIABLE_ERROR_MESSAGE
 
 
-def test_gmsd_zero_for_equal_tensors(prediction: torch.Tensor):
+def test_gmsd_loss_zero_for_equal_tensors(prediction: torch.Tensor, device: str):
```
Comment (snk4tr) [inline tests/test_gmsd.py:99]:
```suggestion
def test_gmsd_loss_zero_for_equal_tensors(prediction: torch.Tensor, device: str) -> None:
```

File: tests/test_gmsd.py [hunk 7]
```diff
@@ -16,160 +18,242 @@ def target() -> torch.Tensor:
...
+    prediction_255 = (prediction * 255).type(torch.uint8)
+    target_255 = (target * 255).type(torch.uint8)
+    measure = gmsd(prediction.to(device), target.to(device))
+
+    measure_255 = gmsd(prediction_255.to(device), target_255.to(device), data_range=255)
+    diff = torch.abs(measure_255 - measure)
+    assert diff <= 1e-4, f'Result for same tensor with different data_range should be the same, got {diff}'
+
+
+def test_gmsd_supports_greyscale_tensors(device: str):
```
Comment (snk4tr) [inline tests/test_gmsd.py:68]:
```suggestion
def test_gmsd_supports_greyscale_tensors(device: str) -> None:
```

File: tests/test_gmsd.py [hunk 8]
```diff
@@ -16,160 +18,242 @@ def target() -> torch.Tensor:
...
 
     loss_255 = GMSDLoss(data_range=255)
-    measure_255 = loss_255(prediction_255, target_255)
+    measure_255 = loss_255(prediction_255.to(device), target_255.to(device))
     diff = torch.abs(measure_255 - measure)
     assert diff <= 1e-4, f'Result for same tensor with different data_range should be the same, got {diff}'
 
 
-def test_gmsd_supports_greyscale_tensors():
+def test_gmsd_loss_supports_greyscale_tensors(device: str):
```
Comment (snk4tr) [inline tests/test_gmsd.py:124]:
```suggestion
def test_gmsd_loss_supports_greyscale_tensors(device: str) -> None:
```

File: tests/test_gmsd.py [hunk 9]
```diff
@@ -16,160 +18,242 @@ def target() -> torch.Tensor:
...
+    loss(prediction.to(device), target.to(device))
 
 
-@pytest.mark.skipif(not torch.cuda.is_available(), reason='No need to run test if there is no GPU.')
-def test_gmsd_supports_greyscale_tensors_on_gpu():
-    loss = GMSDLoss()
-    target = torch.ones(2, 1, 128, 128).cuda()
-    prediction = torch.zeros(2, 1, 128, 128).cuda()
-    loss(prediction, target)
+def test_gmsd_loss_modes(prediction: torch.Tensor, target: torch.Tensor, device: str):
```
Comment (snk4tr) [inline tests/test_gmsd.py:131]:
```suggestion
def test_gmsd_loss_modes(prediction: torch.Tensor, target: torch.Tensor, device: str) -> None:
```

File: tests/test_gmsd.py [hunk 10]
```diff
@@ -16,160 +18,242 @@ def target() -> torch.Tensor:
...
+    multi_scale_gmsd(prediction.to(device), target.to(device), chromatic=True)
 
-@pytest.mark.skipif(not torch.cuda.is_available(), reason='No need to run test if there is no GPU.')
-def test_multi_scale_gmsd_loss_on_gpu(prediction: torch.Tensor, target: torch.Tensor) -> None:
-    prediction = prediction.cuda()
-    target = target.cuda()
-    loss = MultiScaleGMSDLoss(chromatic=True)
-    loss(prediction, target)
+
+def test_multi_scale_gmsd_zero_for_equal_tensors(prediction: torch.Tensor, device: str):
```
Comment (snk4tr) [inline tests/test_gmsd.py:145]:
```suggestion
def test_multi_scale_gmsd_zero_for_equal_tensors(prediction: torch.Tensor, device: str) -> None:
```

File: tests/test_gmsd.py [hunk 11]
```diff
@@ -16,160 +18,242 @@ def target() -> torch.Tensor:
...
+    prediction_255 = (prediction * 255).type(torch.uint8)
+    target_255 = (target * 255).type(torch.uint8)
+
+    measure = multi_scale_gmsd(prediction.to(device), target.to(device))
+    measure_255 = multi_scale_gmsd(prediction_255.to(device), target_255.to(device), data_range=255)
+    diff = torch.abs(measure_255 - measure)
+    assert diff <= 1e-4, f'Result for same tensor with different data_range should be the same, got {diff}'
+
+
+def test_multi_scale_gmsd_supports_greyscale_tensors(device: str):
```
Comment (snk4tr) [inline tests/test_gmsd.py:162]:
```suggestion
def test_multi_scale_gmsd_supports_greyscale_tensors(device: str) -> None:
```

File: tests/test_gmsd.py [hunk 12]
```diff
@@ -16,160 +18,242 @@ def target() -> torch.Tensor:
...
+    target = torch.ones(2, 1, 128, 128)
+    prediction = torch.zeros(2, 1, 128, 128)
+    multi_scale_gmsd(prediction.to(device), target.to(device))
+
+
+def test_multi_scale_gmsd_fails_for_greyscale_tensors_chromatic_flag(device: str):
```
Comment (snk4tr) [inline tests/test_gmsd.py:168]:
```suggestion
def test_multi_scale_gmsd_fails_for_greyscale_tensors_chromatic_flag(device: str) -> None:
```

File: tests/test_gmsd.py [hunk 13]
```diff
@@ -16,160 +18,242 @@ def target() -> torch.Tensor:
...
+    target = torch.ones(2, 1, 128, 128)
+    prediction = torch.zeros(2, 1, 128, 128)
```
Comment (snk4tr) [inline tests/test_gmsd.py:170]: Consider reducing spacial resolution of tensors for the sake of speed here

File: tests/test_gmsd.py [hunk 14]
```diff
@@ -16,160 +18,242 @@ def target() -> torch.Tensor:
...
+
+    measure = multi_scale_gmsd(prediction.to(device), target.to(device))
+    measure_255 = multi_scale_gmsd(prediction_255.to(device), target_255.to(device), data_range=255)
+    diff = torch.abs(measure_255 - measure)
+    assert diff <= 1e-4, f'Result for same tensor with different data_range should be the same, got {diff}'
+
+
+def test_multi_scale_gmsd_supports_greyscale_tensors(device: str):
+    target = torch.ones(2, 1, 128, 128)
+    prediction = torch.zeros(2, 1, 128, 128)
```
Comment (snk4tr) [inline tests/test_gmsd.py:164]: Consider reducing spacial resolution of tensors for the sake of speed here

File: tests/test_gmsd.py [hunk 15]
```diff
@@ -16,160 +18,242 @@ def target() -> torch.Tensor:
...
+    multi_scale_gmsd(prediction.to(device), target.to(device), scale_weights=torch.tensor([3., 4., 2., 1., 2.]))
 
 
-@pytest.mark.skipif(not torch.cuda.is_available(), reason='No need to run test if there is no GPU.')
-def test_multi_scale_gmsd_loss_backward_on_gpu(prediction: torch.Tensor, target: torch.Tensor) -> None:
+def test_multi_scale_gmsd_raise_exception_for_small_images(device: str):
+    target = torch.ones(3, 1, 32, 32)
+    prediction = torch.zeros(3, 1, 32, 32)
+    with pytest.raises(ValueError):
+        multi_scale_gmsd(prediction.to(device), target.to(device), scale_weights=[3., 4., 2., 1., 1.])
```
Comment (snk4tr) [inline tests/test_gmsd.py:184]: Since you are already testing some resolutions that are fine and some resolutions that are too small, could you please estimate the threshold (may be just experimentally?) and put it somewhere in the docsting for the metric and also in the README. It will be super valuable for a user to now this value beforehand

File: tests/test_gmsd.py [hunk 16]
```diff
@@ -16,160 +18,242 @@ def target() -> torch.Tensor:
...
+
+
+def test_multi_scale_gmsd_modes(prediction: torch.Tensor, target: torch.Tensor, device: str):
```
Comment (snk4tr) [inline tests/test_gmsd.py:187]:
```suggestion
def test_multi_scale_gmsd_modes(prediction: torch.Tensor, target: torch.Tensor, device: str) -> None:
```

File: tests/test_gmsd.py [hunk 17]
```diff
@@ -16,160 +18,242 @@ def target() -> torch.Tensor:
...
+    measure = loss(prediction.to(device), target.to(device))
     loss_255 = MultiScaleGMSDLoss(data_range=255)
-    measure_255 = loss_255(prediction_255, target_255)
+    measure_255 = loss_255(prediction_255.to(device), target_255.to(device))
     diff = torch.abs(measure_255 - measure)
     assert diff <= 1e-4, f'Result for same tensor with different data_range should be the same, got {diff}'
 
 
-def test_multi_scale_gmsd_supports_greyscale_tensors():
+def test_multi_scale_gmsd_loss_supports_greyscale_tensors(device: str):
```
Comment (snk4tr) [inline tests/test_gmsd.py:223]:
```suggestion
def test_multi_scale_gmsd_loss_supports_greyscale_tensors(device: str) -> None:
```

File: tests/test_gmsd.py [hunk 18]
```diff
@@ -16,160 +18,242 @@ def target() -> torch.Tensor:
...
-@pytest.mark.skipif(not torch.cuda.is_available(), reason='No need to run test if there is no GPU.')
-def test_multi_scale_gmsd_supports_greyscale_tensors_on_gpu():
-    loss = MultiScaleGMSDLoss()
-    target = torch.ones(2, 1, 128, 128).cuda()
-    prediction = torch.zeros(2, 1, 128, 128).cuda()
-    loss(prediction.cuda(), target.cuda())
-
-
-def test_multi_scale_gmsd_fails_for_greyscale_tensors_chromatic_flag():
+def test_multi_scale_gmsd_loss_fails_for_greyscale_tensors_chromatic_flag(device: str):
```
Comment (snk4tr) [inline tests/test_gmsd.py:230]:
```suggestion
def test_multi_scale_gmsd_loss_fails_for_greyscale_tensors_chromatic_flag(device: str) -> None:
```

File: tests/test_gmsd.py [hunk 19]
```diff
@@ -16,160 +18,242 @@ def target() -> torch.Tensor:
...
 
-@pytest.mark.skipif(not torch.cuda.is_available(), reason='No need to run test if there is no GPU.')
-def test_multi_scale_gmsd_supports_custom_weights_on_gpu(
-        prediction: torch.Tensor, target: torch.Tensor):
-    loss = MultiScaleGMSDLoss(scale_weights=[3., 4., 2., 1., 2.])
-    loss(prediction, target)
-
-
-def test_multi_scale_gmsd_raise_exception_for_small_images():
+def test_multi_scale_gmsd_loss_raise_exception_for_small_images(device: str):
```
Comment (snk4tr) [inline tests/test_gmsd.py:245]:
```suggestion
def test_multi_scale_gmsd_loss_raise_exception_for_small_images(device: str) -> None:
```

File: tests/test_gmsd.py [hunk 20]
```diff
@@ -16,160 +18,242 @@ def target() -> torch.Tensor:
...
     target = torch.ones(3, 1, 32, 32)
     prediction = torch.zeros(3, 1, 32, 32)
     loss = MultiScaleGMSDLoss(scale_weights=[3., 4., 2., 1., 1.])
     with pytest.raises(ValueError):
-        loss(prediction, target)
+        loss(prediction.to(device), target.to(device))
+
+
+def test_multi_scale_loss_gmsd_modes(prediction: torch.Tensor, target: torch.Tensor, device: str):
```
Comment (snk4tr) [inline tests/test_gmsd.py:253]:
```suggestion
def test_multi_scale_loss_gmsd_modes(prediction: torch.Tensor, target: torch.Tensor, device: str) -> None:
```

**Comment by denproc** [comment]
@snk4tr  ready to merge
